### ISSUE TITLE
Rook is refusing to use as OSD an empty and zapped disk partition with error "Used by ceph-disk" in prepare job

### ISSUE THREAD

**Issue by MurzNN** [rook/rook#9923]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

I have an empty and unused by other processes partition on disk, do the zap of it (via dd and other ways too), but `rook-ceph-osd-prepare` job refuses to use it as OSD with those log lines:
```
2022-03-17 09:18:44.969794 D | exec: Running command: udevadm info --query=property /dev/sda4
2022-03-17 09:18:44.973702 D | exec: Running command: lsblk /dev/sda4 --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME
2022-03-17 09:18:44.975058 D | exec: Running command: ceph-volume inventory --format json /dev/sda4
2022-03-17 09:18:45.403744 I | cephosd: skipping device "sda4": ["Used by ceph-disk"].
```

The device `/dev/sda4` is not used by any other processes, here is lsblk of it:
```
# lsblk
NAME         MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda            8:0    0 223.6G  0 disk 
├─sda1         8:1    0     1M  0 part 
├─sda2         8:2    0   977M  0 part /boot
├─sda3         8:3    0    41G  0 part 
│ └─vg0-root 253:0    0  30.5G  0 lvm  /
└─sda4         8:4    0 181.7G  0 part 
sdb            8:16   0 223.6G  0 disk 
```
and here is `fdisk` report:
```
# fdisk -l /dev/sda
Device        Start       End   Sectors   Size Type
/dev/sda1      2048      4095      2048     1M BIOS boot
/dev/sda2      4096   2004991   2000896   977M Linux filesystem
/dev/sda3   2004992  87906303  85901312    41G Linux LVM
/dev/sda4  87906304 468860927 380954624 181.7G Linux filesystem
```

**Expected behavior:**
The `/dev/sda4` should treat as not user and be used as new OSD device.

**How to reproduce it (minimal and precise):**
Maybe try to rebuild the same device structure?

**Environment**:
* OS (e.g. from /etc/os-release): `Ubuntu 20.04`
* Kernel (e.g. `uname -a`): `Linux dgk-m2 5.4.0-100-generic #113-Ubuntu SMP Thu Feb 3 18:43:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`
* Cloud provider or hardware configuration:  `Intel(R) Xeon(R) CPU E5-1650 v4 @ 3.60GHz, 32 GB RAM, 2 x SSD 240 Gb`
* Rook version (use `rook version` inside of a Rook Pod): v1.8.7
* Storage backend version (e.g. for ceph do `ceph -v`): ceph version 16.2.7 (dd0603118f56ab514f133c8d2e3adfc983942503) pacific (stable)
* Kubernetes version (use `kubectl version`): v1.23.4
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Bare Metal
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest/ceph-toolbox.html)): HEALTH_WARN 1 nearfull osd(s); Low space hindering backfill (add storage if this doesn't resolve itself): 2 pgs backfill_toofull; Degraded data redundancy: 2806/632325 objects degraded (0.444%), 1 pg degraded, 5 pgs undersized; 27 pgs not deep-scrubbed in time; 27 pgs not scrubbed in time; 11 pool(s) nearfull

**Comment by MurzNN** [comment]
I've resolved this issue! The source of this problem was the label 'ceph-osd1` that I have attached to that partition using GParted, when was doing initial partitioning!

I've detected this via those log lines in prepare job:
```js
2022-03-17 11:39:45.513289 D | inventory: &{Name:sda4 Parent:sda HasChildren:false DevLinks:/dev/disk/by-partlabel/ceph-osd1 /dev/disk/by-partuuid/18ed3c2f-97c7-4fad-b613-eb0eed89b116 /dev/disk/by-path/pci-0000:00:17.0-ata-1-part4 /dev/disk/by-id/ata-INTEL_SSDSC2BB240G7_BTDV727204X0240AGN-part4 /dev/disk/by-id/wwn-0x55cd2e414dda11bc-part4 Size:195048767488 UUID: Serial:INTEL_SSDSC2BB240G7_BTDV727204X0240AGN Type:part Rotational:false Readonly:false Partitions:[] Filesystem: Vendor: Model:INTEL_SSDSC2BB240G7 WWN:0x55cd2e414dda11bc WWNVendorExtension:0x55cd2e414dda11bc Empty:false CephVolumeData: RealPath:/dev/sda4 KernelName:sda4 Encrypted:false}
2022-03-17 11:39:45.513530 D | cephosd: &{Name:sda4 Parent:sda HasChildren:false DevLinks:/dev/disk/by-partlabel/ceph-osd1 /dev/disk/by-partuuid/18ed3c2f-97c7-4fad-b613-eb0eed89b116 /dev/disk/by-path/pci-0000:00:17.0-ata-1-part4 /dev/disk/by-id/ata-INTEL_SSDSC2BB240G7_BTDV727204X0240AGN-part4 /dev/disk/by-id/wwn-0x55cd2e414dda11bc-part4 Size:195048767488 UUID: Serial:INTEL_SSDSC2BB240G7_BTDV727204X0240AGN Type:part Rotational:false Readonly:false Partitions:[] Filesystem: Vendor: Model:INTEL_SSDSC2BB240G7 WWN:0x55cd2e414dda11bc WWNVendorExtension:0x55cd2e414dda11bc Empty:false CephVolumeData: RealPath:/dev/sda4 KernelName:sda4 Encrypted:false}
```

So the source of the problem was the `/dev/disk/by-partlabel/ceph-osd1` symlink that exists.

The strange thing is that this label is not shown on lsblk:
```
# lsblk -f
NAME         FSTYPE         LABEL UUID                                   FSAVAIL FSUSE% MOUNTPOINT
sda                                                                                     
├─sda1                                                                                  
├─sda2       ext3                 188692fc-7afc-47e3-80bf-a4bde8196d85    818.3M     8% /boot
├─sda3       LVM2_member          lhsOlt-yUxP-ePrh-vuQR-LInI-rqsa-NHKANS                
│ └─vg0-root ext4                 8c8f4429-f874-4dd5-ac6e-c112ae1de04b     10.8G    59% /
└─sda4                                                                                  
sdb          ceph_bluestore                                                             
```

Via `blkid` it also doesn't shown:
```
# blkid -s LABEL -o value /dev/sda4
<empty output>
```
And only via `udev` argument it shows well:
```
# blkid -o udev
ID_FS_PARTLABEL=ceph-osd1
ID_FS_PARTUUID=18ed3c2f-97c7-4fad-b613-eb0eed89b116
```

So I've launched GParted, have removed this label, rebooted the node, and only after this - Ceph starts use this partition as OSD! Hooray!!! :partying_face:

**Comment by cgruver** [comment]
Is this issue not actually a bug?

Is it expected behavior that a storage device cannot be referenced by `/dev/disk/by-partlabel/some-label`?

I have encountered something similar on OpenShift.

I had to delete the label on the partition, and modify my `CephCluster` to directly reference the partition device without prepending it with `/dev/` i.e. I have to reference it as `nvme0n1p5`, not `/dev/nvme0n1p5`...


Should this be re-opened as a bug?  If not, should I open a new issue?

**Comment by travisn** [comment]
The OSDs can be referenced by a path as shown in [this example](https://github.com/rook/rook/blob/release-1.8/deploy/examples/cluster.yaml#L235). 

For raw devices and partitions, they typically just need their names without the `/dev` prefix, so it's expected to use `nvme0n1p5` in that case. @cgruver This is working for you?

**Comment by njspix** [comment]
Just ran into this same issue. AFAICT this is only a problem if 'ceph' is in the label. See:

https://github.com/ceph/ceph/blob/23e12c7096172cfaa9c0f0075ea95d10c7149d94/src/ceph-volume/ceph_volume/util/device.py#L447 

ceph/src/ceph-volume/ceph_volume/util/device.py
```{python}
    def is_ceph_disk_member(self) -> bool:
        def is_member(device: Dict[str, Any]) -> bool:
            return 'ceph' in device.get('PARTLABEL', '') or \
                device.get('PARTTYPE', '') in ceph_disk_guids.keys()
```

**Comment by jobo-prod** [comment]
Ran into the same issue. Imho, this is clearly a bug, although it was deliberately added in the OSD provisioning logic. In my case, I created partitions ("RawVolumes") in Talos Linux, whose names become the partition labels, just like in the issue linked above.

Agnostic from the distro in use, labeling OSD partitions with something like "ceph" is probably not uncommon and having these skipped in osd-prepare is unexpected. What is the reason for this? I don't really get the idea why "ceph" in the label should specifically denote disks already in use.

As a workaround, I relabeled my partitions from `ceph-X` to `cph-X`.

Is it worth reopening this issue @travisn?
